### PR TITLE
add mypy typing

### LINKED ISSUE #5: Add mypy typing
@edthedev @mpitcel @zcarrington since this project is still in its infancy I think it would be wise to add static typing. Typing in Python 3 is optional and not enforced at runtime but does allow for static analysis using [mypy](http://mypy-lang.org/) that can catch bugs that would be hard to find otherwise. I wrote some Python 3 code that uses static typing here:

https://github.com/techservicesillinois/terraform-aws-cloudfront-lambda-basic-auth/blob/master/src/lambda/auth.py

The following link is a useful cheat sheet:

https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html

To actually enforce the typing you would add mypy to your Makefile something like this:

```Makefile
drone: 
	flake8 $(BIN) lambda tests
	mypy lambda tests
	coverage run -m unittest
	coverage report -m --fail-under=100 || (coverage html; exit 1)
```

### PR DESCRIPTION
**Pull request by zdc217**
+ Add mypy typing

Closes #5

### CODE REVIEW

**Review by JPoynton17 — DISMISSED** [review]

**Review by edthedev — DISMISSED** [review]
PR Rejected. I demand a total absence of the benefits of type safety, for traditional reasons.
Kidding. This is awesome. Thanks!

**Review by mpitcel — DISMISSED** [review]

File: app.py [hunk 1]
```diff
@@ -58,7 +61,7 @@ def canary():
     return jsonify({'messages': messages}), status_code
 
 
-def get_clearpass_token(client_id, client_secret):
+def get_clearpass_token(client_id, client_secret) -> str:
```
Comment (ddriddle) [inline app.py]: I recommend you add a type of string to `client_id` and `client_secret`.

File: app.py [hunk 2]
```diff
@@ -8,10 +9,12 @@
...
 
-BASE_URL = os.environ.get('BASE_URL',
-                          'https://cplab1.techservices.illinois.edu:443/api')
-CLIENT_ID = os.environ.get('CLIENT_ID')
-CLIENT_SECRET = os.environ.get('CLIENT_SECRET')
+BASE_URL: str = os.environ.get(
+    'BASE_URL',
+    'https://cplab1.techservices.illinois.edu:443/api'
+)
+CLIENT_ID: Optional[str] = os.environ.get('CLIENT_ID')
```
Comment (ddriddle) [inline app.py]: I recommend you refactor this line as:

```python
CLIENT_ID: str = os.environ['CLIENT_ID']
```

File: app.py [hunk 3]
```diff
@@ -8,10 +9,12 @@
...
+CLIENT_SECRET: Optional[str] = os.environ.get('CLIENT_SECRET')
```
Comment (ddriddle) [inline app.py]: I recommend you refactor as:

```python
CLIENT_SECRET: str = os.environ['CLIENT_SECRET']
```

File: app.py [hunk 4]
```diff
@@ -72,14 +75,14 @@ def get_clearpass_token(client_id, client_secret):
             token_result (dict): Dictionary containing http status code,
                                  token, and error message if applicable.
     """
-    url = f'{BASE_URL}/oauth'
-    data = {
+    url: str = f'{BASE_URL}/oauth'
+    data: Dict[str, Any] = {
```
Comment (ddriddle) [inline app.py]: If you make the changes above you can strength this to ` Dict[str, str]`

**Review by ddriddle — DISMISSED** [review]
I have a few suggestions for strengthening the typing.

File: app.py (hunk 2 above)
Comment (zdc217) [reply app.py]: Thanks, I was struggling with this one. Optional did not seem right.

**Review by MKletz — APPROVED** [review]

**Review by edthedev — APPROVED** [review]

File: app.py (hunk 1 above)
Comment (edthedev) [reply app.py]: I think we actually want something like:
```
def get_clearpass_token(client_id:str, client_secret:str)
```

Comment (zdc217) [reply app.py]: Already added:
image

Comment (zdc217) [reply app.py]: For some reason, Github leaves the old view of the code and marks it "outdated" instead of updating it....

**Pull request merged**